File: src/main/java/org/bbrk24/amurians/EmeryTableScreenHandler.java
package org.bbrk24.amurians;

import com.google.common.collect.Lists;
import java.util.List;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.CraftingResultInventory;
import net.minecraft.inventory.Inventory;
import net.minecraft.inventory.SimpleInventory;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.Property;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.screen.ScreenHandlerContext;
import net.minecraft.screen.ScreenHandlerType;
import net.minecraft.screen.slot.Slot;
import net.minecraft.sound.SoundCategory;
import net.minecraft.sound.SoundEvents;
import net.minecraft.world.World;

// Largely copied from StonecutterScreenHandler, too much is private to subclass it
public class EmeryTableScreenHandler extends ScreenHandler {
    private final ScreenHandlerContext context;
    private final Property selectedRecipe = Property.create();
    private final World world;
    private List<EmeryTableRecipe> availableRecipes = Lists.newArrayList();
    private ItemStack inputStack = ItemStack.EMPTY;
    long lastTakeTime;
    final Slot inputSlot;
    final Slot outputSlot;
    Runnable contentsChangedListener = () -> {};
    public final Inventory input = new SimpleInventory(1) {
        @Override
        public void markDirty() {
            super.markDirty();
            EmeryTableScreenHandler.this.onContentChanged(this);
            EmeryTableScreenHandler.this.contentsChangedListener.run();
        }
    };
    final CraftingResultInventory output = new CraftingResultInventory();

    public EmeryTableScreenHandler(int syncId, PlayerInventory playerInventory) {
        this(syncId, playerInventory, ScreenHandlerContext.EMPTY);
    }

    public EmeryTableScreenHandler(int syncId, PlayerInventory playerInventory, final ScreenHandlerContext context) {
        super(ScreenHandlerType.STONECUTTER, syncId);
        int i;
        this.context = context;
        this.world = playerInventory.player.world;
        this.inputSlot = this.addSlot(new Slot(this.input, 0, 20, 33));
        this.outputSlot = this.addSlot(new Slot(this.output, 1, 143, 33) {
            @Override
            public boolean canInsert(ItemStack stack) {
                return false;
            }

            @Override
            public void onTakeItem(PlayerEntity player, ItemStack stack) {
                stack.onCraft(player.world, player, stack.getCount());
                EmeryTableScreenHandler.this.output.unlockLastRecipe(player);
                ItemStack itemStack = EmeryTableScreenHandler.this.inputSlot.takeStack(1);
                if (!itemStack.isEmpty()) {
                    EmeryTableScreenHandler.this.populateResult();
                }
                context.run((world, pos) -> {
                    long l = world.getTime();
                    if (EmeryTableScreenHandler.this.lastTakeTime != l) {
                        world.playSound(
                            null,
                            pos,
                            SoundEvents.UI_STONECUTTER_TAKE_RESULT,
                            SoundCategory.BLOCKS,
                            1.0f,
                            1.0f
                        );
                        EmeryTableScreenHandler.this.lastTakeTime = l;
                    }
                });
                super.onTakeItem(player, stack);
            }
        });
        for (i = 0; i < 3; ++i) {
            for (int j = 0; j < 9; ++j) {
                this.addSlot(new Slot(playerInventory, j + i * 9 + 9, 8 + j * 18, 84 + i * 18));
            }
        }
        for (i = 0; i < 9; ++i) {
            this.addSlot(new Slot(playerInventory, i, 8 + i * 18, 142));
        }
        this.addProperty(this.selectedRecipe);
    }

    public int getSelectedRecipe() {
        return this.selectedRecipe.get();
    }

    public List<EmeryTableRecipe> getAvailableRecipes() {
        return this.availableRecipes;
    }

    public int getAvailableRecipeCount() {
        return this.availableRecipes.size();
    }

    public boolean canCraft() {
        return this.inputSlot.hasStack() && !this.availableRecipes.isEmpty();
    }

    @Override
    public boolean canUse(PlayerEntity player) {
        return EmeryTableScreenHandler.canUse(this.context, player, Initializer.EMERY_TABLE);
    }

    @Override
    public boolean onButtonClick(PlayerEntity player, int id) {
        if (this.isInBounds(id)) {
            this.selectedRecipe.set(id);
            this.populateResult();
        }
        return true;
    }

    private boolean isInBounds(int id) {
        return id >= 0 && id < this.availableRecipes.size();
    }

    @Override
    public void onContentChanged(Inventory inventory) {
        ItemStack itemStack = this.inputSlot.getStack();
        if (!itemStack.isOf(this.inputStack.getItem())) {
            this.inputStack = itemStack.copy();
            this.updateInput(inventory, itemStack);
        }
    }

    private void updateInput(Inventory input, ItemStack stack) {
        this.availableRecipes.clear();
        this.selectedRecipe.set(-1);
        this.outputSlot.setStack(ItemStack.EMPTY);
        if (!stack.isEmpty()) {
            this.availableRecipes = this.world.getRecipeManager()
                .getAllMatches(Initializer.EMERY_TABLE_RECIPE_TYPE, input, this.world);
        }
    }

    void populateResult() {
        if (!this.availableRecipes.isEmpty() && this.isInBounds(this.selectedRecipe.get())) {
            EmeryTableRecipe stonecuttingRecipe = this.availableRecipes.get(this.selectedRecipe.get());
            this.output.setLastRecipe(stonecuttingRecipe);
            this.outputSlot.setStack(stonecuttingRecipe.craft(this.input));
        } else {
            this.outputSlot.setStack(ItemStack.EMPTY);
        }
        this.sendContentUpdates();
    }

    @Override
    public ScreenHandlerType<EmeryTableScreenHandler> getType() {
        return Initializer.EMERY_TABLE_SCREEN_HANDLER_TYPE;
    }

    public void setContentsChangedListener(Runnable contentsChangedListener) {
        this.contentsChangedListener = contentsChangedListener;
    }

    @Override
    public boolean canInsertIntoSlot(ItemStack stack, Slot slot) {
        return slot.inventory != this.output && super.canInsertIntoSlot(stack, slot);
    }

    @Override
    public ItemStack transferSlot(PlayerEntity player, int index) {
        ItemStack itemStack = ItemStack.EMPTY;
        Slot slot = (Slot) this.slots.get(index);
        if (slot != null && slot.hasStack()) {
            ItemStack itemStack2 = slot.getStack();
            Item item = itemStack2.getItem();
            itemStack = itemStack2.copy();
            if (index == 1) {
                item.onCraft(itemStack2, player.world, player);
                if (!this.insertItem(itemStack2, 2, 38, true)) {
                    return ItemStack.EMPTY;
                }
                slot.onQuickTransfer(itemStack2, itemStack);
            } else if (
                // Like I said, copy-pasted. Not quite legible.
                index == 0 
                    ? !this.insertItem(itemStack2, 2, 38, false)
                    : (
                        this.world.getRecipeManager()
                            .getFirstMatch(
                                Initializer.EMERY_TABLE_RECIPE_TYPE,
                                new SimpleInventory(itemStack2),
                                this.world
                            )
                            .isPresent()
                                ? !this.insertItem(itemStack2, 0, 1, false)
                                : (
                                    index >= 2 && index < 29
                                        ? !this.insertItem(itemStack2, 29, 38, false)
                                        : index >= 29 &&
                                            index < 38 &&
                                            !this.insertItem(itemStack2, 2, 29, false)
                                )
                        )
            ) {
                return ItemStack.EMPTY;
            }
            if (itemStack2.isEmpty()) {
                slot.setStack(ItemStack.EMPTY);
            }
            slot.markDirty();
            if (itemStack2.getCount() == itemStack.getCount()) {
                return ItemStack.EMPTY;
            }
            slot.onTakeItem(player, itemStack2);
            this.sendContentUpdates();
        }
        return itemStack;
    }

    @Override
    public void close(PlayerEntity player) {
        super.close(player);
        this.output.removeStack(1);
        this.context.run((world, pos) -> this.dropInventory(player, this.input));
    }
}

File: src/main/kotlin/org/bbrk24/amurians/EmeryTableScreenHandler.kt
package org.bbrk24.amurians

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.CraftingResultInventory
import net.minecraft.inventory.Inventory
import net.minecraft.inventory.SimpleInventory
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.screen.Property
import net.minecraft.screen.ScreenHandler
import net.minecraft.screen.ScreenHandlerContext
import net.minecraft.screen.slot.Slot
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvents
import net.minecraft.world.World

class EmeryTableScreenHandler(
    syncId: Int,
    playerInventory: PlayerInventory,
    private val context: ScreenHandlerContext
) : ScreenHandler(Initializer.EMERY_TABLE_SCREEN_HANDLER_TYPE, syncId) {
    private val world = playerInventory.player.world
    private val selectedRecipe = Property.create()
    private var inputStack = ItemStack.EMPTY
    private var availableRecipes = mutableListOf<EmeryTableRecipe>()

    var contentsChangedListener = object : Runnable {
        override fun run() { }
    }
    val input = object : SimpleInventory(1) {
        override fun markDirty() {
            super.markDirty()
            onContentChanged(this)
            contentsChangedListener.run()
        }
    }
    val output = CraftingResultInventory()

    private val inputSlot = addSlot(Slot(input, 0, 20, 33))
    private val outputSlot = addSlot(object : Slot(output, 1, 143, 33) {
        override fun canInsert(stack: ItemStack): Boolean = false

        override fun onTakeItem(player: PlayerEntity, stack: ItemStack) {
            Initializer.LOGGER.info("onTakeItem() called")
            stack.onCraft(player.world, player, stack.getCount())
            output.unlockLastRecipe(player)
            val itemStack = inputSlot.takeStack(1)
            if (!itemStack.isEmpty) {
                populateResult()
            }
            context.run { world, pos ->
                val time = world.getTime()
                if (time != lastTakeTime) {
                    world.playSound(
                        null,
                        pos,
                        SoundEvents.UI_STONECUTTER_TAKE_RESULT,
                        SoundCategory.BLOCKS,
                        1.0f,
                        1.0f
                    )
                    lastTakeTime = time
                }
            }
            super.onTakeItem(player, stack)
        }
    })
    private var lastTakeTime = 0L

    init {
        for (i in 1..3) {
            for (j in 0 until 9) {
                addSlot(Slot(playerInventory, j + i * 9, 8 + j * 18, 66 + i * 18));
            }
        }
        for (i in 0 until 9) {
            addSlot(Slot(playerInventory, i, 8 + i * 18, 142));
        }
        addProperty(selectedRecipe);
    }

    constructor(syncId: Int, playerInventory: PlayerInventory) : this(
        syncId,
        playerInventory,
        ScreenHandlerContext.EMPTY
    ) { }

    override fun onContentChanged(inventory: Inventory) {
        val itemStack = inputSlot.getStack()
        if (!itemStack.isOf(inputStack.getItem())) {
            inputStack = itemStack.copy()
            updateInput(inventory, inputStack)
        }
    }

    private fun updateInput(inputInv: Inventory, stack: ItemStack) {
        availableRecipes.clear()
        selectedRecipe.set(-1)
        outputSlot.setStack(ItemStack.EMPTY)
        if (!stack.isEmpty()) {
            availableRecipes = world.getRecipeManager()
                .getAllMatches(Initializer.EMERY_TABLE_RECIPE_TYPE, inputInv, world)
        }
    }

    private fun isInBounds(i: Int): Boolean = i >= 0 && i < availableRecipes.size

    fun populateResult() {
        if (!availableRecipes.isEmpty() && isInBounds(selectedRecipe.get())) {
            val recipe = availableRecipes.get(selectedRecipe.get())
            output.setLastRecipe(recipe)
            outputSlot.setStack(recipe.craft(input))
        } else {
            outputSlot.setStack(ItemStack.EMPTY)
        }
        sendContentUpdates()
    }

    fun getSelectedRecipe(): Int = selectedRecipe.get()
    fun getAvailableRecipes(): List<EmeryTableRecipe> = availableRecipes
    fun getAvailableRecipeCount(): Int = availableRecipes.size
    fun canCraft(): Boolean = inputSlot.hasStack() && !availableRecipes.isEmpty()

    override fun canInsertIntoSlot(unused: ItemStack, slot: Slot) = slot.inventory != output

    override fun canUse(player: PlayerEntity): Boolean {
        return ScreenHandler.canUse(context, player, Initializer.EMERY_TABLE)
    }

    override fun onButtonClick(player: PlayerEntity, id: Int): Boolean {
        if (isInBounds(id)) {
            selectedRecipe.set(id)
            populateResult()
        }
        return true
    }

    public override fun transferSlot(player: PlayerEntity, index: Int): ItemStack {
        var retval = ItemStack.EMPTY
        val slot = slots.get(index) as Slot?
        if (slot != null && slot.hasStack()) {
            val slotStack = slot.getStack()
            retval = slotStack.copy()
            if (index == 1) {
                slotStack.getItem().onCraft(slotStack, player.world, player)
                if (!insertItem(slotStack, 2, 38, true)) {
                    return ItemStack.EMPTY
                }
                slot.onQuickTransfer(slotStack, retval)
            } else if (
                // This part is translated from some decompiled Java code that had a deeply nested
                // ternary expression. I'm not quite sure how it works.
                if (index == 0)
                    !insertItem(slotStack, 2, 38, false)
                else if (
                        world.getRecipeManager()
                            .getFirstMatch(
                                Initializer.EMERY_TABLE_RECIPE_TYPE,
                                SimpleInventory(slotStack),
                                world
                            )
                            .isPresent()
                )
                    !insertItem(slotStack, 0, 1, false)
                else if (index >= 2 && index < 29)
                    !insertItem(slotStack, 29, 38, false)
                else
                    index >= 29 && index < 38 && !insertItem(slotStack, 2, 29, false)
            ) {
                return ItemStack.EMPTY
            }
            if (slotStack.isEmpty()) {
                slot.setStack(ItemStack.EMPTY)
            }
            slot.markDirty()
            if (slotStack.getCount() == retval.getCount()) {
                return ItemStack.EMPTY
            }
            slot.onTakeItem(player, slotStack)
            sendContentUpdates()
        }
        return retval
    }
}
